Add Elasticsearch as an External Logging Service Topic (#951)
astronomer/docs: software/export-task-logs.md

        # needed to prevent zombie deployment worker pods when using KubernetesExecutor
        terminationEndpoint: http://localhost:8000/quitquitquit
    ```
3. Push the configuration change. See [Apply a config change](apply-platform-config.md).

:::info

To revert to the default behavior and export task logs using a Fluentd Daemonset, remove this configuration from your `config.yaml` file.

:::

## Forward Deployment task logs to Elasticsearch

Add your Astronomer Deployment task logs to an existing Elasticsearch instance to centralize log management and analysis. Centralized log management allows you to quickly identify and troubleshoot issues and speeds issue resolution.

### Create an Elastic Deployment and endpoint

1. In your browser, go to https://cloud.elastic.co/ and create a new Deployment. See [Create a deployment](https://www.elastic.co/guide/en/cloud/current/ec-create-deployment.html#ec-create-deployment).
2. Copy and save your deployment credentials when the **Save the deployment credentials** screen appears.
3. On the Elastic dashboard, click the **Gear** icon for your Deployment.

>   ![Elastic Gear icon location](/img/docs/elasticsearch-gear-icon.png)

Author: @jwitz I agree with your concern. However, this icon isn't easy to locate - trust me, I know!

4. Click **Copy endpoint** next to **Elasticsearch**.

    ![Elastic Copy Endpoint location](/img/docs/elasticsearch-copy-endpoint.png)

> 5. Optional. Test the Elasticsearch Deployment endpoint:

Reviewer: Elasticsearch is the (main) component in the Elastic stack. The elastic stack includes Kibana and some other stuff we never use

    - Open a new browser window, paste the endpoint you copied in step 4 in the **Address** bar, and then click **Enter**.
    - Enter the username and password you copied in step 2 and click **Sign in**. Output similar to the following appears:

>     ```text

Author: @jwitz this output confirms the Elasticsearch Deployment endpoint was created successfully. It's an optional step. I'll confirm with the SMEs of it can be removed.

>     ```text

Reviewer: Affirm. Could just include something generic like "if successful, you will be presented with details of the Elastic Cluster."

>     ```text

Author: @stuart23 thanks for confirming. I think that it's obvious that this is an optional step and why it is being performed.

        version	
        number	"8.3.2"
        build_type	"docker"
        build_hash	"8b0b1f23fbebecc3c88e4464319dea8989f374fd"
        build_date	"2022-07-06T15:15:15.901688194Z"
        build_snapshot	false
        lucene_version	"9.2.0"
        minimum_wire_compatibility_version	"7.17.0"
        minimum_index_compatibility_version	"7.0.0"
        tagline	"You Know, for Search"
    ```
6. Run the following command to download the Astronomer Helm chart: 

  ```bash
    helm pull astronomer/astronomer
  ```
7. Run the following command to show the Astronomer Helm chart:

  ```bash
    helm show chart astronomer/astronomer
  ``` 

### Save your Elasticsearch Deployment credentials

After you've created an Elastic Deployment and endpoint, you have two options to save your Elasticsearch Deployment credentials. You can update the `config.yaml` file, or you can create a secret in the Kubernetes cluster.

#### Option 1: Update the `config.yaml` file

1. Run the following command to base64 encode your Elasticsearch Deployment credentials:

  ```bash
    echo -n "<username>:<password>" | base64
  ```
2. Add the following entry to your `config.yaml` file:

>   ```yaml

Author: @jwitz I have done my best to format the code blocks correctly.

        # needed to prevent zombie deployment worker pods when using KubernetesExecutor
        terminationEndpoint: http://localhost:8000/quitquitquit
    ```
3. Push the configuration change. See [Apply a config change](apply-platform-config.md).

:::info

To revert to the default behavior and export task logs using a Fluentd Daemonset, remove this configuration from your `config.yaml` file.

:::

> ## Forward Airflow task logs to Elasticsearch

Reviewer 2: Would rewrite this as "Use an External Elasticsearch for Airflow Task Logs" (but with better English). 

As written, because of existing docs, it sounds like we are sending Airflow Task Logs to the Astronomer ElasticSearch AND external elasticsearch (like we do when forwarding Airflow Task Logs to S3). We are not, we're using the external elasticsearch as a an alternative to the one that ships with Astronomer.

Author: @cmarteepants I have updated the title to: _Use Elasticsearch for Airflow task log management_

Reviewer 2: Better, but we need to say it's an external one to differentiate with the one we ship with. As is, title applies to all scenarios.

Author: @cmarteepants I have updated the title to: _Use an external Elasticsearch instance for Airflow task log management_

> ## Forward Deployment task logs to Elasticsearch

Reviewer 3: @kwcanuck Did you wanna have a prerequisites section here? Or not needed?

> ## Forward Deployment task logs to Elasticsearch

Author: @paolaperaza this is a straightforward process, so there is no need for a prerequisites list.

> Add Airflow task logs from your Astronomer Deployment to an existing Elasticsearch instance on [Elastic Cloud](https://www.elastic.co/cloud/) to centralize log management and analysis. Centralized log management allows you to quickly identify, troubleshoot, and resolve task failure issues.

Reviewer 2: @stuart23 @rob-1126 Curious as to why we decided to make this doc Elastic Cloud specific. We can use any elasticsearch - self-hosted or the cloud managed ones like on AWS, not just the Elastic Cloud one. Maybe we say that this is an example using an Elasticsearch instance on Elastic Cloud?

Author: @cmarteepants we typically document the most common scenario. In this case, it was assumed that Elastic Cloud would be the most common scenario. Also, we're a cloud company and it makes sense to promote the cloud when possible. Ultimately, the procedures should be similar and users should be able to complete them successfully using our content as a reference.

### Create an Elastic Deployment and endpoint

1. In your browser, go to `https://cloud.elastic.co/` and create a new Elastic deployment. See [Create a deployment](https://www.elastic.co/guide/en/cloud/current/ec-create-deployment.html#ec-create-deployment).
2. Copy and save your deployment credentials when the **Save the deployment credentials** screen appears.
3. On the Elastic dashboard, click the **Gear** icon for your Deployment.
  ![Elastic Gear icon location](/img/docs/elasticsearch-gear-icon.png)
4. Click **Copy endpoint** next to **Elasticsearch**.

    ![Elastic Copy Endpoint location](/img/docs/elasticsearch-copy-endpoint.png)

5. Optional. Test the Elasticsearch Deployment endpoint:

>     - Open a new browser window, paste the endpoint you copied in step 4 in the **Address** bar, and then press **Enter**.

Author: @paolaperaza the user doesn't copy the endpoint from the **Address** bar.  It's copied when they select a link on the Elasticsearch dashboard. So, this change is not accurate.

    - Enter the username and password you copied in step 2 and click **Sign in**. Output similar to the following appears:
    ```text
        name	"instance-0000000000"
        cluster_name	"<cluster-name>"
        cluster_uuid	"<cluster-uuid>"
        version	
        number	"8.3.2"
        build_type	"docker"
        build_hash	"8b0b1f23fbebecc3c88e4464319dea8989f374fd"
        build_date	"2022-07-06T15:15:15.901688194Z"
        build_snapshot	false
        lucene_version	"9.2.0"
        minimum_wire_compatibility_version	"7.17.0"
        minimum_index_compatibility_version	"7.0.0"
        tagline	"You Know, for Search"
    ``` 

### Save your Elasticsearch deployment credentials

> After you've created an Elastic deployment and endpoint, you have two options to store your Elastic deployment credentials. You can store the credentials in your Astronomer Software helm values, or as a secret in your Astronomer Software Kubernetes cluster.

Reviewer: The preference question was resolved by Rob.

:::info

To revert to the default behavior and export task logs using a Fluentd Daemonset, remove this configuration from your `config.yaml` file.

:::

## Forward Deployment task logs to Elasticsearch

Add your Astronomer Deployment task logs to an existing Elasticsearch instance to centralize log management and analysis. Centralized log management allows you to quickly identify and troubleshoot issues and resolve issues quickly.

### Create an Elastic Deployment and endpoint

1. In your browser, go to `https://cloud.elastic.co/` and create a new Elastic deployment. See [Create a deployment](https://www.elastic.co/guide/en/cloud/current/ec-create-deployment.html#ec-create-deployment).
2. Copy and save your deployment credentials when the **Save the deployment credentials** screen appears.
3. On the Elastic dashboard, click the **Gear** icon for your Deployment.
  ![Elastic Gear icon location](/img/docs/elasticsearch-gear-icon.png)
4. Click **Copy endpoint** next to **Elasticsearch**.

    ![Elastic Copy Endpoint location](/img/docs/elasticsearch-copy-endpoint.png)

5. Optional. Test the Elasticsearch Deployment endpoint:

>     - Open a new browser window, paste the endpoint you copied in step 4 in the **Address** bar, and then click **Enter**.

Reviewer: ```suggestion
    - Open a new browser window, paste the endpoint you copied in step 4 in the **Address** bar, and then click **Enter**.
    - Open a new browser window, paste the endpoint you copied in step 4 in the **Address** bar to test the connection.
```

Not sure you 'click' enter

Author: @stuart23 if you simply add the endpoint to the Address bar and don't click **Enter**, nothing happens. Clicking **Enter** is a necessary step. To test it for yourself, enter _https://test-for-docs.es.us-central1.gcp.cloud.es.io_ in your browser Address bar and then do nothing. Do you see the authentication prompt? No. If you press **Enter**, you will.

Reviewer: People using this guide have an assumed level of technical competency, so they will know how to open a URL in a browser. 
I was just struggling with the verb 'click', as to me it implies a mouse click. If this is how we should instruct people to visit a URL then that's fine.

Author: @stuart23 thanks for the clarification. Actually, the instruction should be **Press** as this is a keyboard action. I have updated the text. Good catch!

Reviewer: Sorry for the nitpicking! I know you are detail oriented, so I hope you don't mind

Author: @stuart23 not at all. I appreciate your keen eye.

    - Enter the username and password you copied in step 2 and click **Sign in**. Output similar to the following appears:
    ```text
        name	"instance-0000000000"
        cluster_name	"<cluster-name>"
        cluster_uuid	"<cluster-uuid>"
        version	
        number	"8.3.2"
        build_type	"docker"
        build_hash	"8b0b1f23fbebecc3c88e4464319dea8989f374fd"
        build_date	"2022-07-06T15:15:15.901688194Z"
        build_snapshot	false
        lucene_version	"9.2.0"
        minimum_wire_compatibility_version	"7.17.0"
        minimum_index_compatibility_version	"7.0.0"
        tagline	"You Know, for Search"
    ```

> 6. Run the following command to download the Astronomer Helm chart: 

Reviewer: These steps are not needed. I can't remember why I included them. I think it was just to ensure that the chart being used is 0.27+

Author: @stuart23 I assumed that this was a confirmation step and also that you wanted to make sure the chart was downloaded. I have removed the steps as requested.

 ```bash
    helm pull astronomer/astronomer
 ```
7. Run the following command to show the Astronomer Helm chart:

 ```bash
    helm show chart astronomer/astronomer
 ``` 

### Save your Elasticsearch deployment credentials

> After you've created an Elastic deployment and endpoint, you have two options to save your Elastic deployment credentials. You can update your Software `config.yaml` file, or you can create a secret in your Software Kubernetes cluster.

Reviewer: ```suggestion
After you've created an Elastic deployment and endpoint, you have two options to store your Elastic deployment credentials. You can store them directly in your Software helm values, or you can create a secret in your Software Kubernetes cluster.
```

> 

Reviewer: Can we also reference the updating config.yaml instructions here please: https://docs.astronomer.io/software/apply-platform-config

Author: @stuart23 this is a self-contained, simple task. There is no need to link to another topic as the user is unlikely to understand the relevance.

Author: @paolaperaza I added the link, but I think it's unnecessary for the reasons previously stated. If users want or need more information, they can quickly locate it by searching. Also, adding too many links can cause issues when topic titles change, or the topics are removed. The less linking, the easier a documentation set is to manage and maintain. Link rot is a big problem for many companies.

<Tabs
    defaultValue="config.yaml"
    values={[
        {label: 'config.yaml', value: 'config.yaml'},

>         {label: 'Kubernetes cluster', value: 'kubernetes cluster'},

Reviewer: ```suggestion
        {label: 'Kubernetes secret', value: 'kubernetes secret'},
```

    ]}>
<TabItem value="config.yaml">

1. Run the following command to base64 encode your Elasticsearch Deployment credentials:

 ```bash
    echo -n "<username>:<password>" | base64
 ```
2. Add the following entry to your `config.yaml` file:

 ```yaml
 global:
    fluentdEnabled: true
    customLogging:
    enabled: true
    scheme: https
    # host endpoint copied from elasticsearch console with https
    # and port number removed.
    host: “<host-URL>”
    port: "9243"

>     # secret encoded from above step

Reviewer: ```suggestion
    # credentials encoded from above step
```
See comment on next line

>     secret: "<secret>"    

Reviewer: ```suggestion
    secret: "<encoded credentials>"    
```
From someone who has not been exposed to this, is it obvious that <secret> refers to the output of the above step? I think <encoded credentials> might be more straightforward

 ```
3. Add the following entry to your `config.yaml` file to disable internal logging:

 ```yaml
 tags:
   logging: false     
 ```

> 4. Run the following command to upgrade the Astronomer Software release version in the `config.yaml` file:

Reviewer: If we reference the [Apply a Software platform configuration change
](https://docs.astronomer.io/software/apply-platform-config) instructions, we can omit this step as it is already covered

Author: @stuart23 this is a simple step. It's better to provide the necessary information here rather than unnecessarily force a user to link to another topic.

 ```bash
 helm upgrade -f config.yaml --version=0.27 --namespace=<your-platform-namespace> <your-platform-release-name> astronomer/astronomer
 ```

</TabItem>

> <TabItem value="kubernetes cluster">

Reviewer: ```suggestion
<TabItem value="kubernetes secret">
```

1. Run the following command to create a secret for your Elasticsearch Deployment credentials in the Kubernetes cluster:

 ```bash

>  kubectl create secret generic elasticcreds --from-literal elastic=elastic:samplepassword --namespace=<your-platform-namespace>

Reviewer: ```suggestion
 kubectl create secret generic elasticcreds --from-literal elastic=<username>:<password> --namespace=<your-platform-namespace>
```

 ```
2. Add the following entry to your `config.yaml` file:

 ```yaml
 global:
    fluentdEnabled: true
    customLogging:
    enabled: true
    scheme: https
    # host endpoint copied from elasticsearch console with https
    # and port number removed.
    host: “<host-URL>”
    port: "9243"
    # kubernetes secret containing credentials
    secretName: elasticcreds   
 ```
3. Add the following entry to your `config.yaml` file to disable internal logging:

 ```yaml
 tags:
      logging: false    
 ```

> 4. Run the following command to upgrade the Astronomer Software release version in the `config.yaml` file:

Reviewer: Again, we can probably omit this step if we refer to

Author: @stuart23 see my previous comment.
